File: parallel-go/parareal.go
package main

import (
	"fmt"
	"math"
	"os"
	"runtime"
	"sync"
)

// Parameters
const (
	L     = 1.0
	R     = 1.0
	A     = 1.0
	T     = 1.0
	alpha = 0.5
	N     = 1000   // Coarse time steps
	dt    = 0.01   // Coarse time step size
	K     = 2      // Parareal iterations
	M     = 200    // Number of fine solver substeps per coarse step
	chunkSize = 10 // Work batch size for parallel processing
)

// Square wave function
func V(t float64) float64 {
	modTime := math.Mod(t, T)
	if modTime < alpha*T {
		return A
	}
	return 0
}

// RK4 Step for fine solver
func RK4Step(i float64, t float64, dt float64) float64 {
	f := func(t, i float64) float64 { return (V(t) - R*i) / L }

	k1 := dt * f(t, i)
	k2 := dt * f(t+dt/2, i+k1/2)
	k3 := dt * f(t+dt/2, i+k2/2)
	k4 := dt * f(t+dt, i+k3)

	return i + (k1+2*k2+2*k3+k4)/6
}

// Full RK4 solution
func SolveRK4(filename string) {
	file, _ := os.Create(filename)
	defer file.Close()

	i := 0.0
	t := 0.0
  dt := 1e-4
  N := int(1e5) 
	for step := 0; step < N; step++ {
		fmt.Fprintf(file, "%f %f\n", t, i)
		i = RK4Step(i, t, dt)
		t += dt
	}
}

// Parareal Algorithm
func Parareal(filename string) {
	runtime.GOMAXPROCS(runtime.NumCPU())

	var wg sync.WaitGroup
	var mu sync.Mutex

	pool := sync.Pool{
		New: func() interface{} { return make([]float64, N+1) },
	}
	i0 := pool.Get().([]float64)
	i1 := pool.Get().([]float64)
	t := pool.Get().([]float64)
	defer pool.Put(i0)
	defer pool.Put(i1)
	defer pool.Put(t)

	for step := 0; step <= N; step++ {
		t[step] = float64(step) * dt
	}

	// Initial coarse solution (Euler method)
	for step := 0; step < N; step++ {
		i0[step+1] = i0[step] + dt*(V(t[step])-R*i0[step])/L
	}

	// Parareal Iterations
	for k := 0; k < K; k++ {
		wg.Add(N / chunkSize)
		for step := 0; step < N; step += chunkSize {
			go func(start int) {
				defer wg.Done()
				for j := start; j < start+chunkSize && j < N; j++ {
					fine := i0[j]
					fine_dt := dt / float64(M)
					for m := 0; m < M; m++ {
						fine = RK4Step(fine, t[j]+float64(m)*fine_dt, fine_dt)
					}

					mu.Lock()
					i1[j+1] = fine + i0[j+1] - i0[j] 
					mu.Unlock()
				}
			}(step)
		}
		wg.Wait()
		copy(i0, i1)
	}

	file, _ := os.Create(filename)
	defer file.Close()
	for step := 0; step <= N; step++ {
		fmt.Fprintf(file, "%f %f\n", t[step], i0[step])
	}
}

func main() {
	SolveRK4("rk4.txt")
	Parareal("parareal.txt")
}

File: parallel-go/parareal.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp
from concurrent.futures import ProcessPoolExecutor
import timeit
# Parameters
L = 1.0       # Inductance (H)
R = 1.0       # Resistance (Ohm)
A = 1.0       # Amplitude of square wave
T = 2.0       # Time period of square wave
alpha = 0.5   # Duty ratio (fraction of time V=A)

def V(t):
    return A if (t % T) < (alpha * T) else 0

def dI_dt(t, i):
    return (V(t) - R * i) / L

def coarse_solver(t0, t1, i0, N_coarse):
    t_values = np.linspace(t0, t1, N_coarse + 1)
    dt = (t1 - t0) / N_coarse
    i_values = np.zeros(len(t_values))
    i_values[0] = i0
    for n in range(N_coarse):
        i_values[n+1] = i_values[n] + dt * dI_dt(t_values[n], i_values[n])
    return i_values[-1]

def fine_solver(t0, t1, i0, N_fine):
    t_values = np.linspace(t0, t1, N_fine + 1)
    sol = solve_ivp(dI_dt, [t0, t1], [i0], t_eval=t_values, method="RK45")
    return sol.y[0][-1]

def parareal_parallel(t0, t_final, i0, N_steps, N_coarse, N_fine, max_iter=10):
    t_points = np.linspace(t0, t_final, N_steps + 1)
    dt = (t_final - t0) / N_steps
    i_coarse = np.zeros(N_steps + 1)
    i_fine = np.zeros(N_steps + 1)
    i_corrected = np.zeros(N_steps + 1)

    i_coarse[0] = i_corrected[0] = i0
    for n in range(N_steps):
        i_coarse[n+1] = coarse_solver(t_points[n], t_points[n+1], i_corrected[n], N_coarse)
        i_corrected[n+1] = i_coarse[n+1]

    for k in range(max_iter):
        i_fine[0] = i0
        with ProcessPoolExecutor() as executor:
            futures = [
                executor.submit(fine_solver, t_points[n], t_points[n+1], i_corrected[n], N_fine)
                for n in range(N_steps)
            ]
            fine_results = [f.result() for f in futures]

        for n in range(N_steps):
            i_fine[n+1] = fine_results[n]
        
        max_error = 0
        for n in range(N_steps):
            correction = i_fine[n+1] - i_coarse[n+1]
            i_corrected[n+1] += correction
            max_error = max(max_error, abs(correction))

        print(f"Iteration {k+1}, Max Error: {max_error:.6f}")
        if max_error < 1e-6:
            break
    
    return t_points, i_corrected

def reference_rk4_solution(t0, t_final, i0, N_ref):
    t_values = np.linspace(t0, t_final, N_ref + 1)
    sol = solve_ivp(dI_dt, [t0, t_final], [i0], t_eval=t_values, method="RK45")
    return sol.t, sol.y[0]

t0, t_final = 10, 50
N_steps = 1000     # Parareal time steps
N_coarse = 500     # Steps in coarse solver
N_fine = 5000      # Steps in fine solver
max_iter = 1    # Max iterations for convergence
N_ref = 500000      # High-resolution RK4 for reference

t_parareal, i_parareal = parareal_parallel(t0, t_final, i0=0, N_steps=N_steps, N_coarse=N_coarse, N_fine=N_fine, max_iter=max_iter)
t_rk4, i_rk4 = reference_rk4_solution(t0, t_final, 0, N_ref)
print(timeit.timeit(lambda: parareal_parallel(t0, t_final, i0=0, N_steps=N_steps, N_coarse=N_coarse, N_fine=N_fine, max_iter=max_iter), number=1))
print(timeit.timeit(lambda: reference_rk4_solution(t0, t_final, 0, N_ref), number=1))
plt.figure(figsize=(8,5))
plt.plot(t_parareal, i_parareal, 'b-', label="Parareal Approximation (Parallel)")
plt.plot(t_rk4, i_rk4, 'r-', label="Reference RK4 Solution")
plt.xlabel("Time (s)")
plt.ylabel("Current (A)")
plt.title("Parareal Algorithm vs Reference RK4 Solution")
plt.legend()
plt.grid()
plt.show()

